### PR TITLE
LS336 only, develop build

### PR DESCRIPTION
**Pull request by zatkins2**
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] Unless I am preparing a release, I have opened this PR onto the `develop` branch.

### CODE REVIEW

**Comment by zatkins2** [comment]
Only the 'transport' feature of Operation 'servo_to_temperature' is not tested, for cryogenic limitation reasons. Everything else works.

**Comment by zatkins2** [comment]
Debugged transport feature of temperature servoing.

**Comment by zatkins2** [comment]
Added a 'transport_offset' optional kwarg to temperature servoing to allow transporting load to setpoint - transport_offset if raising the temperature.

File: agents/lakeshore336/LS336_agent.py
Comment (BrianJKoopman) [inline agents/lakeshore336/LS336_agent.py:655]: I'd reword this to something like "Gets an arbitrary channel attribute and stores it in the `session.data` dict." Right now it sounds like your getting the attribute from the `session.data` dict.

Please also add an example for the structure of the data stored within `session.data` for reference in the documentation. See the 372 Agent acquisition process for an example: https://github.com/simonsobs/socs/blob/b6dde01c3f66dc4dcc528262de7428ba19d88b95/agents/lakeshore372/LS372_agent.py#L182

File: socs/Lakeshore/Lakeshore336.py
Comment (BrianJKoopman) [inline socs/Lakeshore/Lakeshore336.py:290]: I think you might be trying to avoid the built-in function `input` here? Convention is to put an underscore after the variable, a leading underscore would imply a private variable. Would change this to `input_`

File: socs/Lakeshore/Lakeshore336.py
Comment (BrianJKoopman) [inline socs/Lakeshore/Lakeshore336.py:443]: This looks largely unmodified. If this can be the same class as in the 372 you should import it from the 372 driver. You could use that 372 Curve as a base class and extend it with the one added method here.

File: socs/Lakeshore/Lakeshore336.py
Comment (BrianJKoopman) [inline socs/Lakeshore/Lakeshore336.py:1183]: I find it most useful to give the option here to connect to any device given an argument. That allows one to run this file interactively and perform some actions with a Lakeshore device, so like in the 372 it initializes `ls = LS372(sys.argv[1])`, which allows me to run something like `python3 -i Lakeshore372.py 192.168.1.2` and get the 372 at that address, and interact as needed.

I'd recommend something similar here for consistency with the other Lakeshore drivers, which will include using `ls` instead of `box` and leaving the comm open.

**Review by BrianJKoopman — CHANGES_REQUESTED** [review]
Most of the comments can be found inline here. There's a few big things, like the use of the locking mechanism. The current implementation might appear to work in testing, but I don't think it's respecting the lock in certain circumstances, so a serial collision seems possible to me.

Some general things:
* Documentation is needed. See the other Agent documentation files in `socs/docs/agents/` for examples. Configuration examples are a must.
* It would be useful to add the most recent data points from the `acq` process to the `session.data` object. That makes them easily to query from the client. See https://ocs.readthedocs.io/en/develop/developer/session-data.html
* PEP08 recommendations. There's a lot of spacing and other formatting issues that go against the convention in the rest of OCS, which is to follow PEP08. You can get a full list by running `python3 -m flake8 LS336_agent.py` (`pip3 install flake8` first if you don't have flake8 already.) I made some specific comments related to these, but that'll catch the repeats/others I didn't comment on.

File: agents/lakeshore336/Dockerfile
Comment (zatkins2) [reply agents/lakeshore336/Dockerfile:14]: done

File: agents/lakeshore336/LS336_agent.py
Comment (zatkins2) [reply agents/lakeshore336/LS336_agent.py:48]: done

File: agents/lakeshore336/LS336_agent.py
Comment (zatkins2) [reply agents/lakeshore336/LS336_agent.py:47]: This is interestingly on lists of top 10 python noob mistakes.....learn something new everyday...fixed

File: agents/lakeshore336/LS336_agent.py
Comment (zatkins2) [reply agents/lakeshore336/LS336_agent.py:82]: fixed

File: agents/lakeshore336/LS336_agent.py
Comment (zatkins2) [reply agents/lakeshore336/LS336_agent.py:518]: fixed, this was caught by flake -- FYI, another cool benefit of VS Code is you can autoformat a file according to PEP08 with one click. It fixes most things, but not all (like this).

File: socs/Lakeshore/Lakeshore336.py
Comment (zatkins2) [reply socs/Lakeshore/Lakeshore336.py:12]: indeed, removed

File: socs/Lakeshore/Lakeshore336.py
Comment (zatkins2) [reply socs/Lakeshore/Lakeshore336.py:109]: done, i have also increased this to 0.1 seconds so as not to live on the edge (manual specs 50ms)

File: socs/Lakeshore/Lakeshore336.py
Comment (zatkins2) [reply socs/Lakeshore/Lakeshore336.py:156]: you are 100% right, good catch, fixed

File: socs/Lakeshore/Lakeshore336.py
Comment (zatkins2) [reply socs/Lakeshore/Lakeshore336.py:195]: just in case someone wanted it for some reason. i agree i cannot immediately think of a use-case.  so as not to confuse people, i am removing it.

File: socs/Lakeshore/Lakeshore336.py
Comment (zatkins2) [reply socs/Lakeshore/Lakeshore336.py:129]: 1. I have added an assert statement to the msg method to ensure the delay is greater than 50ms for all messages, including queries
2. This will be true by the above (since 1/0.05 = 20)

1. This is fixed, see previous conversation
2. This is true by default in the pyserial read_until method (it looks for '\n', and LS336 terminates messages with '\r\n')
3. See above
4. See above

File: agents/lakeshore336/LS336_agent.py
Comment (zatkins2) [reply agents/lakeshore336/LS336_agent.py:160]: The warn/continue logic has been replaced by a return False statement

File: agents/lakeshore336/LS336_agent.py
Comment (zatkins2) [reply agents/lakeshore336/LS336_agent.py:565]: No reason, I think I moved these all to the driver Lakeshore336.py, should be cleaner now.

File: agents/lakeshore336/LS336_agent.py
Comment (zatkins2) [reply agents/lakeshore336/LS336_agent.py:571]: This Task can take a while to run; it's cryogenically limited (eg, around 5min say). As with acq, have replaced this with a return False statement.

Comment (zatkins2) [reply agents/lakeshore336/LS336_agent.py:571]: The code I have should be functionally the same as the release_and_acquire method, except with an added sleep statement between the release and the attempted acquisition. This I think better reflects expected behavior of such a process (same thing for acq): it only attempts to acquire the lock when it needs it (e.g. for reading a temperature, infrequently) and the lock is otherwise "totally free" in the interim for short processes. Since the act of reading a temperature is short, short processes can have short timeouts (e.g. 3 seconds) without having to worry that the release_and_acquire method is called frequently-enough in acq or here (eg, with a dummy variable "last_release" etc).

Comment (zatkins2) [reply agents/lakeshore336/LS336_agent.py:571]: *Short tasks can have short timeouts

File: agents/lakeshore336/LS336_agent.py
Comment (zatkins2) [reply agents/lakeshore336/LS336_agent.py:578]: This is a misnomer: it's not a bug, more just something to be careful of.

When ramping, the device takes control of the setpoint, moving it from its old value to the commanded value. If we want to check temperature stability against the commanded value, we need to store that value separate from the current machine setpoint.

File: agents/lakeshore336/LS336_agent.py
Comment (zatkins2) [reply agents/lakeshore336/LS336_agent.py:650]: Hm I think this is benign -- it makes it consistent with all other Tasks, which have a call to "wait." I agree it's probably unnecessary, but it is nice to give Operations breathing room.

Note in the case that temperatures are stable, the method doesn't reach the call to wait.

I'll leave it here (and in all other methods) unless you'd like me to remove them.

File: agents/lakeshore336/LS336_agent.py
Comment (zatkins2) [reply agents/lakeshore336/LS336_agent.py:647]: agreed, removed

File: agents/lakeshore336/LS336_agent.py
Comment (zatkins2) [reply agents/lakeshore336/LS336_agent.py:655]: 1. done
2. done

File: socs/Lakeshore/Lakeshore336.py
Comment (zatkins2) [reply socs/Lakeshore/Lakeshore336.py:290]: yes, fixed and fixed all others too

File: socs/Lakeshore/Lakeshore336.py
Comment (zatkins2) [reply socs/Lakeshore/Lakeshore336.py:1183]: done

File: socs/Lakeshore/Lakeshore336.py
Comment (BrianJKoopman) [inline socs/Lakeshore/Lakeshore336.py:151]: I like the idea here, but if we encounter this it'll raise an AssertionError and stop the program. Maybe better would be to force delays shorter than 50 ms to be 50 ms and then log a warning about the provided delay. This ensures the msg still goes through with proper timing and the user is notified about their improper delay.

(We didn't make this a parameter in other Lakeshore drivers because this timing *has* to be respected, else communication isn't guaranteed to work properly.)

File: socs/Lakeshore/Lakeshore336.py
Comment (BrianJKoopman) [reply socs/Lakeshore/Lakeshore336.py]: re: 4. -- maybe `read_until` used to be used, but the call to `self.com.recv()` doesn't ensure the entire response is read, so we should make sure it fits in 4096 bytes. Everything else looks good.

**Comment by BrianJKoopman** [comment]
Coming back to this PR. @zhuber21 has interest in this Agent at Cornell, and I know @kaiwen-zheng131 has been using the Lakeshore336 hardware at Princeton. It would be great if the two of you could coordinate on any final updates to this agent (I went through and marked old conversation as resolved if the latest commits did so), and rebase on the latest development branch before we merge the Agent. This might require a new branch be made based on this on. Please get in touch if there are git organizational issues I can help with.

File: agents/lakeshore336/LS336_agent.py
Comment (BrianJKoopman) [reply agents/lakeshore336/LS336_agent.py:713]: I still think we could do without these and leave it to the client to wait longer if it's really needed.

File: agents/lakeshore336/LS336_agent.py
Comment (zhuber21) [reply agents/lakeshore336/LS336_agent.py:16]: @BrianJKoopman Do you think it is worth also worth adding all the class attributes here like the M1000 agent does? There are a lot of them, though several of them are just repeats of the parameters (which I have added to the docstring).

File: agents/lakeshore336/LS336_agent.py
Comment (zhuber21) [reply agents/lakeshore336/LS336_agent.py:713]: I will remove the parameter from all the methods then. There is currently a 100 ms delay baked into LS336.msg(). We can make it longer or add specific delays for specific commands if it becomes necessary in testing.

File: socs/Lakeshore/Lakeshore336.py
Comment (zhuber21) [reply socs/Lakeshore/Lakeshore336.py:401]: Compared to the Curve class in Lakeshore372.py, it looks like there are some small differences besides the addition of the new soft_cal() function. Functions like _set_header, set_format (only in the docstring), and _set_data_point differ due to different parameters in CRVHDR and CRVPT between the 372 and 336. set_curve() has an extra 1 second delay added after the curve is uploaded with the comment "necessary to make work." The get_header() function call within delete_curve() is commented out here but isn't in the 372 code.

Comment (BrianJKoopman) [reply socs/Lakeshore/Lakeshore336.py:401]: Thanks for the info, going to mark as resolved, since we're going to leave this for now.

**Review by BrianJKoopman — APPROVED** [review]
Apologies for letting the latest changes sit here for a while. I just pushed some fixes for some formatting in the documentation that was generating warnings in the build. I'll squash and merge once checks complete. The Lakeshore336Agent docker images should become available shortly after on Docker Hub.

**Pull request merged**
